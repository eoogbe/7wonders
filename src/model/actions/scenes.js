/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const VIEW_WONDER_CHOOSER = 'VIEW_WONDER_CHOOSER';

export const VIEW_WONDER_PANEL = 'VIEW_WONDER_PANEL';

export const VIEW_GAME_BOARD = 'VIEW_GAME_BOARD';

export const START_AGE = 'START_AGE';

export const SCORE_GAME = 'SCORE_GAME';

export const QUIT = 'QUIT';

export const viewWonderChooser = (playerId) => ({
  type: VIEW_WONDER_CHOOSER,
  payload: {
    playerId,
  },
});

export const viewWonderPanel = () => ({
  type: VIEW_WONDER_PANEL,
});

export const viewGameBoard = () => ({
  type: VIEW_GAME_BOARD,
});

export const startAge = () => ({
  type: START_AGE,
});

export const scoreGame = () => ({
  type: SCORE_GAME,
});

export const quit = () => ({
  type: QUIT,
});
