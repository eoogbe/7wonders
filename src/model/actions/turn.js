/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const BUILD_STRUCTURE = 'BUILD_STRUCTURE';

export const BUILD_WONDER = 'BUILD_WONDER';

export const ADD_MONEY = 'ADD_MONEY';

export const SELECT_HAND_CARD = 'SELECT_HAND_CARD';

export const CHOOSE_NEIGHBOR = 'CHOOSE_NEIGHBOR';

export const COPY_GUILD = 'COPY_GUILD';

export const buildStructure = (cardId) => ({
  type: BUILD_STRUCTURE,
  payload: {
    cardId,
  },
});

export const buildFreeStructure = () => ({
  type: BUILD_STRUCTURE,
  payload: {
    isFree: true,
  },
});

export const buildWonder = () => ({
  type: BUILD_WONDER,
});

export const addMoney = () => ({
  type: ADD_MONEY,
});

export const selectHandCard = (cardId, cardIdx) => ({
  type: SELECT_HAND_CARD,
  payload: {
    cardId,
    cardIdx,
  },
});

export const chooseNeighbor = (costType) => ({
  type: CHOOSE_NEIGHBOR,
  payload: {
    costType,
  },
});

export const copyGuild = (cardId) => ({
  type: COPY_GUILD,
  payload: {
    cardId,
  },
});
