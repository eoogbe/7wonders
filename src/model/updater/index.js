/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { concat, decorate, withDefaultState } from 'redux-fp';
import ages from './ages';
import app, { initialAppState } from './app';
import copyAiGuild from './copyAiGuild';
import hands from './hands';
import military from './military';
import neighbors from './neighbors';
import players from './players';
import takeAiTurns from './takeAiTurns';
import trade from './trade';
import wonder from './wonder';

const initialState = {
  ages: [],
  players: {},
  app: initialAppState,
};

export default decorate(
  withDefaultState(initialState),
  concat(
    wonder,
    neighbors,
    takeAiTurns,
    ages,
    hands,
    players,
    military,
    copyAiGuild,
    app,
    trade
  )
);
