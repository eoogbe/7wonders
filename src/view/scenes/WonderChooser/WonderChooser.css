/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.WonderChooser {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  position: relative;
  height: 100%;
}

.WonderChooser-title {
  font-size: 1.5rem;
  margin: 0;
}

.WonderChooser-selectedWonder {
  font-size: 1rem;
  display: flex;
  align-items: flex-start;
  margin: 1.5em 0;
}

.WonderChooser-stageList {
  display: flex;
}

.WonderChooser-wonderSide {
  font-size: 1.25rem;
  margin-left: 0.5em;
  padding: 0.125em 0.5em;
  border-radius: 0.75em;
  outline: none;
  color: #FFF;
  cursor: pointer;
}

.WonderChooser-wonderList {
  font-size: 1rem;
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  width: 55em;
}

.WonderChooser-wonder {
  font-size: 1rem;
  margin: 0.25em;
  padding: 0.25em;
  border: 1px solid rgba(93, 64, 55);
  cursor: pointer;
}

.WonderChooser-wonder.selected {
  border: 2px solid #000;
}

.WonderChooser-nextButton {
  font-size: 1.5rem;
  margin-top: 1em;
}
