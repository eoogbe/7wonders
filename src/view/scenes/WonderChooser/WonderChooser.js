/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import classnames from 'classnames';
import { compose } from 'ramda';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from 'model/actions';
import {
  getWonderChooserWonderSide, getWonderChooserStages, getWonderChooserWonders,
} from 'presentation/wonder';
import QuitButton from 'view/components/QuitButton';
import * as theme from 'view/theme';
import './WonderChooser.css';
import WonderStage from './WonderStage';

const renderStage = (stage, idx) => (
  <WonderStage key={idx} {...stage} />
);

const renderWonder = ({ selectWonder }) => ({ id, name, isSelected }) => (
  <li
    key={id}
    className={classnames('WonderChooser-wonder', { selected: isSelected })}
    onClick={() => selectWonder(id)}
  >
    {name}
  </li>
);

const WonderChooser = () => {
  const selectedStages = useSelector(getWonderChooserStages);
  const wonderSide = useSelector(getWonderChooserWonderSide);
  const wonders = useSelector(getWonderChooserWonders);

  const dispatch = useDispatch();
  const selectWonder = compose(dispatch, actions.selectWonder);
  const toggleWonderSide = compose(dispatch, actions.toggleWonderSide);
  const viewWonderPanel = compose(dispatch, actions.viewWonderPanel);

  const wonderSideStyle = { backgroundColor: theme.primaryColor };
  const quitButtonStyle = {
    position: 'absolute',
    left: 0,
    bottom: 0,
  };

  return (
    <section className="WonderChooser">
      <h1 className="WonderChooser-title">
        Choose your wonder:
      </h1>
      <div className="WonderChooser-selectedWonder">
        <ol className="WonderChooser-stageList">
          {selectedStages.map(renderStage)}
        </ol>
        <button
          className="WonderChooser-wonderSide"
          style={wonderSideStyle}
          type="button"
          onClick={toggleWonderSide}
        >
          {wonderSide}
        </button>
      </div>
      <ul className="WonderChooser-wonderList">
        {wonders.map(renderWonder({ selectWonder }))}
      </ul>
      <button
        className="WonderChooser-nextButton primary-button"
        type="button"
        onClick={viewWonderPanel}
      >
        Choose wonder
      </button>
      <QuitButton style={quitButtonStyle} />
    </section>
  );
};

export default WonderChooser;
