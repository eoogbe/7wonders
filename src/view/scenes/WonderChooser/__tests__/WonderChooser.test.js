/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import PLAYERS from '__fixtures__/players';
import { selectWonder, toggleWonderSide, viewWonderPanel } from 'model/actions';
import WonderChooser from '../WonderChooser';

const mockStore = configureStore();

it('renders without crashing', () => {
  const store = mockStore(createState({
    wonder: 'PYRAMIDS_OF_GIZA',
  }));
  const tree =
    renderer.create(
      <Provider store={store}>
        <WonderChooser />
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('selects the wonder', () => {
  const store = mockStore(createState({
    wonder: 'PYRAMIDS_OF_GIZA',
  }));
  const tree = mount(
    <Provider store={store}>
      <WonderChooser />
    </Provider>
  );
  tree.find('.WonderChooser-wonder').first().simulate('click');
  expect(store.getActions()).toEqual([selectWonder('COLOSSUS_OF_RHODES')]);
});

it('toggles the wonder side', () => {
  const store = mockStore(createState({
    wonder: 'PYRAMIDS_OF_GIZA',
  }));
  const tree = mount(
    <Provider store={store}>
      <WonderChooser />
    </Provider>
  );
  tree.find('.WonderChooser-wonderSide').simulate('click');
  expect(store.getActions()).toEqual([toggleWonderSide()]);
});

it('views the wonder panel', () => {
  const store = mockStore(createState({
    wonder: 'PYRAMIDS_OF_GIZA',
  }));
  const tree = mount(
    <Provider store={store}>
      <WonderChooser />
    </Provider>
  );
  tree.find('.WonderChooser-nextButton').simulate('click');
  expect(store.getActions()).toEqual([viewWonderPanel()]);
});

const createState = (player) => ({
  app: {
    currentPlayer: 'PLAYER1',
    wonderChooserPlayer: 'PLAYER1',
  },
  players: {
    ...PLAYERS,
    PLAYER1: {
      ...PLAYERS.PLAYER1,
      ...player,
    },
  },
});
