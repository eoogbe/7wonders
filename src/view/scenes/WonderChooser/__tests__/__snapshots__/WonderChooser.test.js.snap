// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<section
  className="WonderChooser"
>
  <h1
    className="WonderChooser-title"
  >
    Choose your wonder:
  </h1>
  <div
    className="WonderChooser-selectedWonder"
  >
    <ol
      className="WonderChooser-stageList"
    >
      <li
        className="WonderChooserWonderStage"
      >
        <ul
          className="WonderChooserWonderStage-cost"
        >
          <li>
            Stone
          </li>
          <li>
            Stone
          </li>
        </ul>
        <ul
          className="WonderChooserWonderStage-effectList"
        >
          <li>
            3 points
          </li>
        </ul>
      </li>
      <li
        className="WonderChooserWonderStage"
      >
        <ul
          className="WonderChooserWonderStage-cost"
        >
          <li>
            Wood
          </li>
          <li>
            Wood
          </li>
          <li>
            Wood
          </li>
        </ul>
        <ul
          className="WonderChooserWonderStage-effectList"
        >
          <li>
            5 points
          </li>
        </ul>
      </li>
      <li
        className="WonderChooserWonderStage"
      >
        <ul
          className="WonderChooserWonderStage-cost"
        >
          <li>
            Stone
          </li>
          <li>
            Stone
          </li>
          <li>
            Stone
          </li>
          <li>
            Stone
          </li>
        </ul>
        <ul
          className="WonderChooserWonderStage-effectList"
        >
          <li>
            7 points
          </li>
        </ul>
      </li>
    </ol>
    <button
      className="WonderChooser-wonderSide"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(93, 64, 55)",
        }
      }
      type="button"
    >
      A
    </button>
  </div>
  <ul
    className="WonderChooser-wonderList"
  >
    <li
      className="WonderChooser-wonder"
      onClick={[Function]}
    >
      Colossus of Rhodes
    </li>
    <li
      className="WonderChooser-wonder"
      onClick={[Function]}
    >
      Hanging garden of Babylon
    </li>
    <li
      className="WonderChooser-wonder"
      onClick={[Function]}
    >
      Lighthouse of Alexandria
    </li>
    <li
      className="WonderChooser-wonder"
      onClick={[Function]}
    >
      Mausoleum of Halicarnassus
    </li>
    <li
      className="WonderChooser-wonder selected"
      onClick={[Function]}
    >
      Pyramids of Giza
    </li>
    <li
      className="WonderChooser-wonder"
      onClick={[Function]}
    >
      Statue of Zeus in Olympia
    </li>
    <li
      className="WonderChooser-wonder"
      onClick={[Function]}
    >
      Temple of Artemis in Ephesus
    </li>
  </ul>
  <button
    className="WonderChooser-nextButton primary-button"
    onClick={[Function]}
    type="button"
  >
    Choose wonder
  </button>
  <button
    className="QuitButton"
    onClick={[Function]}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
      }
    }
    type="button"
  >
    Quit
  </button>
</section>
`;
